consensus/safety-rules/src/safety_rules.rs

```diff
@@ -89,17 +97,47 @@ impl SafetyRules {
         }
     }
 
+    /// This reconciles the key pair of a validator signer with a given validator set
+    /// during epoch changes.
+    /// Current impl of consensus does not expect key reconciliation to panic or fail.
+    /// Instead, we signal debug messages indicating following error causes:
+    ///     1. Validator not in the set
+    ///     2. Validator in the set, but no matching key found in storage
+    fn reconcile_key(&mut self, epoch_state: &EpochState) -> Result<(), Error> {
+        if let Some(expected_key) = epoch_state.verifier.get_public_key(&self.author) {
+            let curr_key = self.signer().ok().map(|s| s.public_key());
+            if curr_key != Some(expected_key.clone()) {
+                debug!(
+                    "Reconciled pub key for signer {} [{:#?} -> {}]",
+                    self.author, curr_key, expected_key
+                );
+                let consensus_key = self
+                    .persistent_storage
+                    .consensus_key_for_version(expected_key)
+                    .expect("Unable to retrieve consensus private key");
```

Comment:
This isn't production worthy code.

1) Operator starts safety rules and lets it run for a long time
2) Operator over-rotates and now the key for the next round isn't actually in secure storage
3) Operator then fixes issue

The node will always crash when seeing it isn't in the set, even though there's not much that we can do.

Alternatively consider the case where
1) SafetyRules is nice and far into the future
2) There was a corruption of state and we need to start from a time before our current keys but that will lead up to our current keys
3) As we replay the updates, safety rules is going to get a set of keys that they no longer have access to

This should be an error and not a panic, this is recoverable. Please adjust.

cc: @zekun000

Reply:
also to be clear, if we do not find the key, we should set the signer to none.

Reply:
Fixed. Now it returns an Error::InternalError{"Validator key not found"} instead of just panic()